core/src/main/java/org/opensearch/sql/data/model/ExprValueUtils.java

```diff
@@ -177,4 +181,51 @@ public static Map<String, ExprValue> getTupleValue(ExprValue exprValue) {
   public static Boolean getBooleanValue(ExprValue exprValue) {
     return exprValue.booleanValue();
   }
+
+  /**
+   * Convert a datetime value to milliseconds since Epoch.
+   * @param value A value.
+   * @return Milliseconds since Epoch.
+   */
+  public static long extractEpochMilliFromAnyDateTimeType(ExprValue value) {
+    switch ((ExprCoreType)value.type()) {
+      case DATE:
+        return MILLIS.between(LocalDateTime.of(1970, 1, 1, 0, 0), value.datetimeValue());
```

Comment:
Nice catch!
Initially `case DATE:` fallen back with `TIMESTAMP` and `DATETIME`, conversion thru timestamp (`Instant`) returned wrong result, because timestamp changes TZ to UTC.
`DATE` -> `MILLIS` -> `DATE` -> `MILLIS` check failed.

Reply:
Thank you. While thinking on your comment I found a mistake. Fixed in 478141e7e.